kmod/src/data.c

```diff
@@ -772,8 +773,13 @@ static int scoutfs_write_begin(struct file *file,
 	if (ret == 0)
 		ret = block_write_begin(mapping, pos, len, flags, pagep,
 					scoutfs_get_block_write);
-	if (ret)
+	if (ret) {
 		scoutfs_release_trans(sb);
+		if (ret == -ENOBUFS) {
+			scoutfs_inc_counter(sb, alloc_trans_retry);
+			goto retry;
```

Comment:
We need to unlock the inode index item locks before retrying in a new transaction.  The inode index items specifically index inodes by the transaction they were modified in.  We need to re-acquire the locks for the new transaction number.  (But in this loop we only want to retry if the errno comes from block_write_begin, so it might take a bit of surgery with error paths and gotos and such to make sense.)

Reply:
ok cool. Fun.